amworkflow/occ_helpers.py

```diff
@@ -191,3 +206,245 @@ def topo_explorer(shape: TopoDS_Shape, shape_type: str) -> list:
+    """
+    @brief Returns the topo shape intersecting the item at the given position.
+    @param position Position of the plane in world coordinates.
+    @param axis Axis along which of the direction.
+    @return TopoDS_Shape with intersection or empty TopoDS_Shape if no intersection is found.
+    """
+    intsctr = BRepAlgoAPI_Common
+    xmin, ymin, zmin, xmax, ymax, zmax = get_occ_bounding_box(item)
+    bnd_x = abs(xmin - xmax) * 1.2
+    bnd_y = abs(ymin - ymax) * 1.2
+    bnd_z = abs(zmin - zmax) * 1.2
+    match axis:
+        case "z":
+            plan_len = max(bnd_x, bnd_y)
+            p1, v1 = gp_Pnt(0, 0, position), gp_Vec(0, 0, 1)
+            fc1 = make_face(
+                gp_Pln(p1, vec_to_dir(v1)), -plan_len, plan_len, -plan_len, plan_len
+            )
+    common = intsctr(item, fc1)
+    common.Build()
+
+
+def scale(item: TopoDS_Shape, cnt_pnt: gp_Pnt, factor: float) -> TopoDS_Shape:
+    """
+    @brief Scales TopoDS_Shape to a given value. This is useful for scaling shapes that are in a shape with respect to another shape.
+    @param item TopoDS_Shape to be scaled.
+    @param cnt_pnt the point of the scaling center.
+    @param factor Factor to scale the shape by. Default is 1.
+    @return a scaled TopoDS_Shape with scaling applied to it.
+    """
+    scaling_transform = gp_Trsf()
+    scaling_transform.SetScale(cnt_pnt, factor)
+    scaled_shape = BRepBuilderAPI_Transform(item, scaling_transform, True).Shape()
+    return scaled_shape
+
+
+def carve_hollow(face: TopoDS_Shape, factor: float):
+    """
+    @brief (This can be replaced by cutter3D() now.)Carving on a face with a shape scaling down from itself.
+    @param face TopoDS_Shape to be cutted.
+    @param factor Factor to be used to scale the cutter.
+    @return A shape with the cutter in it's center of mass scaled by factor
+    """
+    cnt = get_face_center_of_mass(face, gp_pnt=True)
+    cutter = scale(face, cnt, factor)
+    cut = BRepAlgoAPI_Cut(face, cutter).Shape()
+    return cut
+
+
+def rotate_face(shape: TopoDS_Shape, angle: float, axis: str = "z", cnt: tuple = None):
+    """
+    @brief Rotate the topography by the given angle around the center of mass of the face.
+    @param shape TopoDS_Shape to be rotated.
+    @param angle Angle ( in degrees ) to rotate by.
+    @param axis determine the rotation axis.
+    @return the rotated shape.
+    """
+    transform = gp_Trsf()
+    if cnt is None:
+        cnt = get_face_center_of_mass(shape, gp_pnt=True)
+    match axis:
+        case "z":
+            ax = gp_Ax1(cnt, gp_Dir(0, 0, 1))
+        case "y":
+            ax = gp_Ax1(cnt, gp_Dir(0, 1, 0))
+        case "x":
+            ax = gp_Ax1(cnt, gp_Dir(1, 0, 0))
+    transform.SetRotation(ax, angle)
+    transformed = BRepBuilderAPI_Transform(shape, transform).Shape()
+    return transformed
+
+
+def fuse(shape1: TopoDS_Shape, shape2: TopoDS_Shape) -> TopoDS_Shape:
```

Comment:
Do we really need that as an own function? It's just calling the OCC function

Reply:
I suggest to keep this kind of simple helpers. Although it is just a simple wrapper, to remember to which the function it wraps belongs is not very easy. For example, this function **BRepAlgoAPI_Fuse** comes from **OCC.Core.BRepAlgoAPI**. However there are dozens of classes share the same or similar segements like "Brep", "Algo". I agree that it would be a good practice to try to remember raw functions from OCC if you aim to master it but if not, my thought is that to simply wrap it up with an easier name and interface would be saving a lot of work. 

Moreover, if you look into these raw functions, for example this  **BRepAlgoAPI_Fuse**:

```python
"""
        Empty constructor.

        Returns
        -------
        None
        Empty constructor <pf> - pavefiller object that is carried out.

        Parameters
        ----------
        PF: BOPAlgo_PaveFiller

        Returns
        -------
        None
        Constructor with two shapes <s1> -argument <s2> -tool <anoperation> - the type of the operation obsolete.

        Parameters
        ----------
        S1: TopoDS_Shape
        S2: TopoDS_Shape
        theRange: Message_ProgressRange,optional
        	default value is Message_ProgressRange()

        Returns
        -------
        None
        Constructor with two shapes <s1> -argument <s2> -tool <anoperation> - the type of the operation <pf> - pavefiller object that is carried out obsolete.

        Parameters
        ----------
        S1: TopoDS_Shape
        S2: TopoDS_Shape
        aDSF: BOPAlgo_PaveFiller
        theRange: Message_ProgressRange,optional
        	default value is Message_ProgressRange()

        Returns
        -------
        None

        """
```
You would see that it does not only take S1 and S2 but could be a lot of strange parameters. This is common in OCC that it offers verbose settings for one simple function. Since we would not have a chance to use most of them, I think it would be easier to hide them behind the helper. Do you think it more or less make sense? That is what I thought but I can certainly delete them if you does not feel right about it.